File: dclib/core/src/dclib/eventing/EventDelegate.java
package dclib.eventing;

import java.util.ArrayList;
import java.util.Collection;

public final class EventDelegate<T> {
	
	private final Collection<T> listeners = new ArrayList<T>();
	
	public final void listen(final T listener) {
    	if (!listeners.contains(listener)) {
    		listeners.add(listener);
        }
	}
	
	public final void notify(final Event<T> event) {
        for (T listener : listeners) {
        	event.notify(listener);
        }
	}
	
	public final void remove(final T listener) {
		listeners.remove(listener);
	}

}

File: dclib/core/src/main/java/dclib/eventing/EventDelegate.kt
package dclib.eventing

class EventDelegate<T> {
    var handlers = listOf<(T) -> Unit>()

    infix fun on(handler: (T) -> Unit) {
        handlers += handler
    }
	
	infix fun remove(handler: (T) -> Unit) {
		handlers -= handler
	}

    fun notify(event: T) {
        for (handler in handlers) {
            handler(event)
        }
    }
}